hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -131,21 +134,46 @@ public void beforeInitialize(CollectionPersister persister, int anticipatedSize)
 	}
 
 	@Override
+	@SuppressWarnings( "unchecked" )
 	public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
 		final Type elementType = persister.getElementType();
-		final List sn = (List) getSnapshot();
+		final List<Object> sn = (List<Object>) getSnapshot();
 		if ( sn.size() != bag.size() ) {
 			return false;
 		}
-		for ( Object elt : bag ) {
-			final boolean unequal = countOccurrences( elt, bag, elementType ) != countOccurrences( elt, sn, elementType );
-			if ( unequal ) {
+
+		// HHH-11032 - Group objects by Type.getHashCode() to reduce the complexity of the search
+		final Map<Integer, List<Object>> hashToInstancesBag = groupByEqualityHash( bag, elementType );
+		final Map<Integer, List<Object>> HashToInstancesSn = groupByEqualityHash( sn, elementType );
+		if ( hashToInstancesBag.size() != HashToInstancesSn.size() ) {
+			return false;
+		}
+
+		for ( Map.Entry<Integer, List<Object>> hashToInstancesBagEntry : hashToInstancesBag.entrySet() ) {
```

Comment:
it might be worth iterating the entrySet twice: first just check all of `instancesBag.size() != instancesSn.size()` match. The only if they really do match, you do the loop to check the elements.

Reply:
I'm thinking about this one. The only time this would make a difference is if there are:
 
1) the same element multiple times;
2) many elements with the same hash.

1) could obviously happen for an ```ElementCollection``` or ```@ManyToMany`` `, but I don't know how often it really happens in practice, or if it could result in many or just a few duplicates. For a one-to-many association, it should never happen. This is actually a good area for optimization. I'll address that in a general comment.

2) probably shouldn't happen all that much; if there are few elements with the same hash, and many elements in the bag, it might be costly to do the extra iteration.

Reply:
ok but (2) is actually quite likely if the objects are indeed different: you'd immediately spot some (likely) change in hash distribution.

Of course iterating twice is slightly more expensive than iterating only once, so the case in which all objects are the same would be slightly slower - I suppose not much though as each loop iteration is simpler.

But if there's any difference, you'd spot it much earlier as apparently it's the `isSame` operation which is slow - you'd be skipping _all of them_ which is quite great :)

Reply:
```isSame``` will depend on the element type.

For most types, ```isSame``` delegates to ```isEqual``, which again depends on the type.

For ```EntityType```, ```isSame``` returns ```x == y```.

For ```ComponentType```, ```isSame``` first checks if ```x == y``. If ```false```, ```isSame``` is called for each sub-property.

IOW, calling ```isSame``` for entity elements will be a lot faster than for ```String``` elements.

Reply:
@Sanne, I just want to make sure you see that ```iSame``` performance is highly type-dependent...

Reply:
Sure, I understand and agree.  A consequence is that it might be that `isSame` is very cheap, but it might be expensive.

The comment on https://hibernate.atlassian.net/browse/HHH-11032 seems to suggest that - at least in the use case of the original issue reporter - dodging invoking the `isSame` method as much as possible would be good:

> Since isSame uses reflection, it's quite slow.

(I'm not sure to what reflection specifically he refers, but I guess either way `isSame` is indeed more expensive than the other methods, at least because it's highly type dependent)

Reply:
For associations, ```isSame``` is very cheap; it only involves ```x == y```. I think it is worthwhile to do things differently for associations vs basic/embeddable elements. WDYT?

Reply:
interesting. Ok, send a PR and I'll review it :)

But let's not over think this? At some point we need to balance the fact to reach a quick fix for those who need it, which as far as we know were happy enough already with the original patch. 

We can always improve things in subsequent patches later - and possibly keep a simpler approach for the stable backports, while tinkering with something more aggressive for master?

So far we know:
 - the "old algorythm" allocates less so it should be preferred for small bags (let's say small ~ 10 ?)
 - otherwise use the new strategy which uses buckets by hash, iterate only once (like you said) and yet try to exit the second loop early on (when the count exceeds the expected figure).

I'm not sure how to take advantage of the fact that associations are cheaper, I guess I'll see your proposal :)

Reply:
@Sanne, I'll update the PR tomorrow.

Reply:
@Sanne, I realized that even for non-associations, ```is_same``` will probably mostly be doing reference checks.

I'll make this change as well.